src/main/kotlin/com/github/julianhille/sqldelight/jscipherdriver/JsSqlCipherDriver.kt

```diff
@@ -0,0 +1,207 @@
+    this.transaction = transaction
+    if (enclosing == null) {
+      db.exec("BEGIN TRANSACTION")
+    }
+    return transaction
+  }
+
+  override fun currentTransaction() = transaction
+
+  private inner class Transaction(
+    override val enclosingTransaction: Transacter.Transaction?
+  ) : Transacter.Transaction() {
+    override fun endTransaction(successful: Boolean) {
+      if (enclosingTransaction == null) {
+        if (successful) {
+          db.exec("END TRANSACTION")
+        } else {
+          db.exec("ROLLBACK TRANSACTION")
+        }
+      }
+      transaction = enclosingTransaction
+    }
+  }
+
+  private fun createOrGetStatement(identifier: Int?, sql: String): Statement = db.prepare(sql)
+
+  fun getVersion(): Int = (db.pragma("user_version") as? Double)?.toInt() ?: 0
+
+  private fun setVersion(version: Int): Unit {
+    db.pragma("user_version=${version}")
+  }
+
+  fun migrateIfNeeded(
+    create: (SqlDriver) -> Unit,
+    upgrade: (SqlDriver, Int, Int) -> Unit,
+    version: Int
+  ) {
+    val initialVersion = getVersion()
+    println("soimething mumble mumble $initialVersion")
+    println("soimething mumble mumble $db")
+    println("soimething mumble mumble $transaction")
+    val driver = this
+    if (initialVersion == 0) {
+      transaction.run {
+        create(driver)
+        setVersion(version)
+      }
+    } else if (initialVersion != version) {
+      if (initialVersion > version) {
+          throw IllegalStateException("Database version $initialVersion newer than config version $version")
+        }
+        transaction.run {
+          upgrade(driver, initialVersion, version)
+          setVersion(version)
+        }
+    }
+  }
+}
+
+class SqlJsCipherStatement: SqlPreparedStatement {
+  val parameters = mutableListOf<Any?>()
+  override fun bindBytes(index: Int, bytes: ByteArray?) {
+    parameters.add(bytes)
+  }
+
+  override fun bindLong(index: Int, long: Long?) {
+    // We convert Long to Double because Kotlin's Double is mapped to JS number
+    // whereas Kotlin's Long is implemented as a JS object
+    parameters.add(long?.toDouble())
+  }
+
+  override fun bindDouble(index: Int, double: Double?) {
+    parameters.add(double)
+  }
+
+  override fun bindString(index: Int, string: String?) {
+    parameters.add(string)
+  }
+
+  fun bindBoolean(index: Int, boolean: Boolean?) {
+    parameters.add(
+      when (boolean) {
+        null -> null
+        true -> 1.0
+        false -> 0.0
+      }
+    )
+  }
+}
+
+class SqlJsCipherCursor(private val statementIterator: StatementIterator): SqlCursor {
+  var lastResult: IteratorResult? = null
+  var columns: List<String> = listOf<String>()
+
+
+  private fun getIndex(index: Int):  dynamic? /* Number | String | Uint8Array | Nothing? */ {
+    val name = columns[index]
+    val value = lastResult?.value ?: throw NoSuchElementException()
+    return js("value[name]")
+  }
+
+  override fun close() {
+    statementIterator.`return`()
```

Comment:
No `freemem` ?

https://github.com/cashapp/sqldelight/blob/cb699fcc19632a70deeb2930470bcf09db625a42/drivers/sqljs-driver/src/main/kotlin/app/cash/sqldelight/driver/sqljs/JsSqlDriver.kt#L137

Reply:
https://github.com/sql-js/sql.js/blob/ec44c18a421ea2d9934acbc54fc12b31b35635eb/src/api.js#L657 exists in  sql js but does not and is not needed in https://github.com/m4heshd/better-sqlite3-multiple-ciphers/blob/master/docs/api.md#class-statement